// @lobehub/tts/es/react/useSpeechRecognition/useSpeechRecognitionCore.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import { useCallback, useEffect, useState } from 'react';
import { SpeechRecognition } from "../../core/const/polyfill";
export var useSpeechRecognitionCore = function useSpeechRecognitionCore(locale) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
    onTextChange = _ref.onTextChange,
    onRecognitionStart = _ref.onRecognitionStart,
    onRecognitionFinish = _ref.onRecognitionFinish,
    onRecognitionStop = _ref.onRecognitionStop,
    onRecognitionError = _ref.onRecognitionError;
  var _useState = useState(null),
    _useState2 = _slicedToArray(_useState, 2),
    recognition = _useState2[0],
    setRecognition = _useState2[1];
  var _useState3 = useState(''),
    _useState4 = _slicedToArray(_useState3, 2),
    text = _useState4[0],
    setText = _useState4[1];
  var _useState5 = useState(false),
    _useState6 = _slicedToArray(_useState5, 2),
    isLoading = _useState6[0],
    setIsLoading = _useState6[1];
  var _useState7 = useState(false),
    _useState8 = _slicedToArray(_useState7, 2),
    isFinalStop = _useState8[0],
    setFinalStop = _useState8[1];
  useEffect(function () {
    if (recognition) return;
    try {
      var speechRecognition = new SpeechRecognition();
      speechRecognition.interimResults = true;
      speechRecognition.continuous = true;
      speechRecognition.onstart = function () {
        setFinalStop(false);
        setIsLoading(true);
      };
      speechRecognition.onend = function () {
        setIsLoading(false);
        setFinalStop(true);
      };
      speechRecognition.onresult = function (_ref2) {
        var _result$;
        var results = _ref2.results;
        if (!results) return;
        var result = results[0];
        if (!isFinalStop && result !== null && result !== void 0 && (_result$ = result[0]) !== null && _result$ !== void 0 && _result$.transcript) {
          var _value = result[0].transcript;
          setText(_value);
          onTextChange === null || onTextChange === void 0 || onTextChange(_value);
        }
        if (result.isFinal) {
          speechRecognition.abort();
        }
      };
      setRecognition(speechRecognition);
    } catch (error) {
      console.error('Error useSpeechRecognitionCore:', error);
      onRecognitionError === null || onRecognitionError === void 0 || onRecognitionError(error);
    }
  }, [isFinalStop]);
  useEffect(function () {
    if (!isLoading && text) {
      onRecognitionFinish === null || onRecognitionFinish === void 0 || onRecognitionFinish(text);
    }
  }, [text, isLoading]);
  useEffect(function () {
    if (recognition) recognition.lang = locale;
  }, [recognition, locale]);
  var handleStart = useCallback(function () {
    setText('');
    onTextChange === null || onTextChange === void 0 || onTextChange('');
    try {
      recognition.start();
      onRecognitionStart === null || onRecognitionStart === void 0 || onRecognitionStart();
    } catch (error) {
      console.error('Error useSpeechRecognitionCore:', 'start', error);
      onRecognitionError === null || onRecognitionError === void 0 || onRecognitionError(error);
    }
  }, [recognition]);
  var handleStop = useCallback(function () {
    try {
      recognition.abort();
      onRecognitionStop === null || onRecognitionStop === void 0 || onRecognitionStop();
    } catch (error) {
      console.error('Error useSpeechRecognitionCore:', 'stop', error);
      onRecognitionError === null || onRecognitionError === void 0 || onRecognitionError(error);
    }
    setIsLoading(false);
  }, [recognition]);
  return {
    isLoading: isLoading,
    start: handleStart,
    stop: handleStop,
    text: text
  };
};
