// @lobehub/ui/es/Button/style.js
var _templateObject, _templateObject2, _templateObject3, _templateObject4;
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
import { createStyles } from 'antd-style';
export var useStyles = createStyles(function (_ref) {
  var css = _ref.css,
    token = _ref.token,
    stylish = _ref.stylish,
    prefixCls = _ref.prefixCls;
  return {
    filled: css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n      &.", "-btn.", "-btn-variant-filled {\n        ", ";\n      }\n    "])), prefixCls, prefixCls, stylish.variantFilled),
    glass: stylish.blur,
    outlined: css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n      &.", "-btn.", "-btn-variant-outlined {\n        ", "\n      }\n    "])), prefixCls, prefixCls, stylish.variantOutlined),
    root: css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n      &.", "-btn {\n        transition:\n          color 400ms ", ",\n          background 100ms ", ";\n\n        > .", "-btn-icon {\n          display: flex;\n        }\n      }\n    "])), prefixCls, token.motionEaseOut, token.motionEaseOut, prefixCls),
    shadow: css(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n      box-shadow: ", " !important;\n    "])), token.boxShadowTertiary)
  };
});
