// @lobehub/ui/es/mobile/MobileTabBar/index.js
'use client';

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import { memo } from 'react';
import { Flexbox } from 'react-layout-kit';
import useControlledState from 'use-merge-value';
import MobileSafeArea from "../MobileSafeArea";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var MobileTabBar = /*#__PURE__*/memo(function (_ref) {
  var className = _ref.className,
    safeArea = _ref.safeArea,
    style = _ref.style,
    items = _ref.items,
    activeKey = _ref.activeKey,
    defaultActiveKey = _ref.defaultActiveKey,
    onChange = _ref.onChange;
  var _useControlledState = useControlledState(defaultActiveKey || items[0].key, {
      defaultValue: defaultActiveKey,
      onChange: onChange,
      value: activeKey
    }),
    _useControlledState2 = _slicedToArray(_useControlledState, 2),
    currentActive = _useControlledState2[0],
    setCurrentActive = _useControlledState2[1];
  var _useStyles = useStyles(),
    styles = _useStyles.styles,
    cx = _useStyles.cx;
  return /*#__PURE__*/_jsxs(Flexbox, {
    className: cx(styles.container, className),
    style: style,
    children: [/*#__PURE__*/_jsx(Flexbox, {
      align: 'center',
      className: styles.inner,
      horizontal: true,
      justify: 'space-around',
      children: items.map(function (item) {
        var active = item.key === currentActive;
        return /*#__PURE__*/_jsxs(Flexbox, {
          align: 'center',
          className: cx(styles.tab, active && styles.active),
          gap: 4,
          justify: 'center',
          onClick: function onClick() {
            var _item$onClick;
            setCurrentActive(item.key);
            item === null || item === void 0 || (_item$onClick = item.onClick) === null || _item$onClick === void 0 || _item$onClick.call(item);
          },
          children: [/*#__PURE__*/_jsx(Flexbox, {
            align: 'center',
            className: styles.icon,
            justify: 'center',
            children: typeof item.icon === 'function' ? item.icon(active) : item.icon
          }), /*#__PURE__*/_jsx("div", {
            className: styles.title,
            children: typeof item.title === 'function' ? item.title(active) : item.title
          })]
        }, item.key);
      })
    }), safeArea && /*#__PURE__*/_jsx(MobileSafeArea, {
      position: 'bottom'
    })]
  });
});
export default MobileTabBar;
