// @lobehub/ui/es/Mermaid/index.js
'use client';

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _excluded = ["children", "copyButtonSize", "fullFeatured", "copyable", "showLanguage", "style", "type", "enablePanZoom", "className", "bodyRender", "actionsRender"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import { isUndefined } from 'lodash-es';
import { memo } from 'react';
import { Flexbox } from 'react-layout-kit';
import CopyButton from "../CopyButton";
import Tag from "../Tag";
import FullFeatured from "./FullFeatured";
import MermaidContainer from "./components/MermaidContainer";
import MermaidZoomableContainer from "./components/MermaidZoomableContainer";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var Mermaid = /*#__PURE__*/memo(function (_ref) {
  var children = _ref.children,
    _ref$copyButtonSize = _ref.copyButtonSize,
    copyButtonSize = _ref$copyButtonSize === void 0 ? 'site' : _ref$copyButtonSize,
    fullFeatured = _ref.fullFeatured,
    _ref$copyable = _ref.copyable,
    copyable = _ref$copyable === void 0 ? true : _ref$copyable,
    _ref$showLanguage = _ref.showLanguage,
    showLanguage = _ref$showLanguage === void 0 ? true : _ref$showLanguage,
    style = _ref.style,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'block' : _ref$type,
    enablePanZoom = _ref.enablePanZoom,
    className = _ref.className,
    bodyRender = _ref.bodyRender,
    actionsRender = _ref.actionsRender,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useStyles = useStyles(type),
    cx = _useStyles.cx,
    styles = _useStyles.styles;
  var tirmedChildren = children.trim();
  var originalActions = copyable && /*#__PURE__*/_jsx(CopyButton, {
    content: tirmedChildren,
    placement: "left",
    size: copyButtonSize
  });
  var actions = actionsRender ? actionsRender({
    actionIconSize: copyButtonSize,
    content: children,
    originalNode: originalActions
  }) : originalActions;
  var defaultBody = enablePanZoom || isUndefined(enablePanZoom) && fullFeatured ? /*#__PURE__*/_jsx(MermaidZoomableContainer, {
    children: tirmedChildren
  }) : /*#__PURE__*/_jsx(MermaidContainer, {
    children: tirmedChildren
  });
  var body = bodyRender ? bodyRender({
    content: tirmedChildren,
    originalNode: defaultBody
  }) : defaultBody;
  if (fullFeatured) return /*#__PURE__*/_jsx(FullFeatured, _objectSpread(_objectSpread({
    actionsRender: actionsRender,
    bodyRender: bodyRender,
    className: className,
    content: tirmedChildren,
    copyable: copyable,
    showLanguage: showLanguage,
    style: style,
    type: type
  }, rest), {}, {
    children: body
  }));
  return /*#__PURE__*/_jsxs("div", _objectSpread(_objectSpread({
    className: cx(styles.container, className),
    "data-code-type": "mermaid",
    style: style
  }, rest), {}, {
    children: [/*#__PURE__*/_jsx(Flexbox, {
      align: 'center',
      className: styles.button,
      flex: 'none',
      gap: 4,
      horizontal: true,
      children: actions
    }), !enablePanZoom && showLanguage && /*#__PURE__*/_jsx(Tag, {
      className: styles.lang,
      children: "Mermaid"
    }), /*#__PURE__*/_jsx("div", {
      className: styles.scroller,
      children: body
    })]
  }));
});
export default Mermaid;
