// @lobehub/ui/es/Highlighter/Highlighter.js
'use client';

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _excluded = ["fullFeatured", "actionIconSize", "children", "language", "className", "copyable", "showLanguage", "variant", "shadow", "wrap", "bodyRender", "actionsRender", "enableTransformer", "theme"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';
import { Flexbox } from 'react-layout-kit';
import CopyButton from "../CopyButton";
import Tag from "../Tag";
import FullFeatured from "./FullFeatured";
import SyntaxHighlighter from "./SyntaxHighlighter";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
export var Highlighter = /*#__PURE__*/memo(function (_ref) {
  var fullFeatured = _ref.fullFeatured,
    actionIconSize = _ref.actionIconSize,
    children = _ref.children,
    _ref$language = _ref.language,
    language = _ref$language === void 0 ? 'markdown' : _ref$language,
    className = _ref.className,
    _ref$copyable = _ref.copyable,
    copyable = _ref$copyable === void 0 ? true : _ref$copyable,
    _ref$showLanguage = _ref.showLanguage,
    showLanguage = _ref$showLanguage === void 0 ? true : _ref$showLanguage,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'filled' : _ref$variant,
    shadow = _ref.shadow,
    wrap = _ref.wrap,
    bodyRender = _ref.bodyRender,
    actionsRender = _ref.actionsRender,
    enableTransformer = _ref.enableTransformer,
    theme = _ref.theme,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useStyles = useStyles(),
    styles = _useStyles.styles,
    cx = _useStyles.cx;
  var variants = useMemo(function () {
    return cva(styles.root, {
      defaultVariants: {
        shadow: false,
        variant: 'filled',
        wrap: false
      },
      /* eslint-disable sort-keys-fix/sort-keys-fix */
      variants: {
        variant: {
          filled: styles.filled,
          outlined: styles.outlined,
          borderless: styles.borderless
        },
        shadow: {
          false: null,
          true: styles.shadow
        },
        wrap: {
          false: styles.nowrap,
          true: styles.wrap
        }
      }
      /* eslint-enable sort-keys-fix/sort-keys-fix */
    });
  }, [styles]);
  var tirmedChildren = children.trim();
  var originalActions = copyable && /*#__PURE__*/_jsx(CopyButton, {
    content: tirmedChildren,
    glass: true,
    size: actionIconSize
  });
  var actions = actionsRender ? actionsRender({
    actionIconSize: actionIconSize,
    content: tirmedChildren,
    language: language,
    originalNode: originalActions
  }) : originalActions;
  var originalBody = /*#__PURE__*/_jsx(SyntaxHighlighter, {
    enableTransformer: enableTransformer,
    language: language === null || language === void 0 ? void 0 : language.toLowerCase(),
    theme: theme,
    variant: variant,
    children: tirmedChildren
  });
  var body = bodyRender ? bodyRender({
    content: tirmedChildren,
    language: language,
    originalNode: originalBody
  }) : originalBody;
  if (fullFeatured) return /*#__PURE__*/_jsx(FullFeatured, _objectSpread(_objectSpread({
    actionsRender: actionsRender,
    className: className,
    content: tirmedChildren,
    copyable: copyable,
    language: language,
    shadow: shadow,
    showLanguage: showLanguage,
    variant: variant,
    wrap: wrap
  }, rest), {}, {
    children: body
  }));
  return /*#__PURE__*/_jsxs(Flexbox, _objectSpread(_objectSpread({
    className: cx(variants({
      shadow: shadow,
      variant: variant,
      wrap: wrap
    }), className),
    "data-code-type": "highlighter"
  }, rest), {}, {
    children: [/*#__PURE__*/_jsx(Flexbox, {
      align: 'center',
      className: styles.actions,
      flex: 'none',
      gap: 4,
      horizontal: true,
      children: actions
    }), showLanguage && language && /*#__PURE__*/_jsx(Tag, {
      className: styles.lang,
      children: language.toLowerCase()
    }), body]
  }));
});
Highlighter.displayName = 'Highlighter';
export default Highlighter;
