// @lobehub/ui/es/Image/index.js
'use client';

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _excluded = ["wrapperClassName", "style", "preview", "isLoading", "minSize", "size", "actions", "alwaysShowActions", "objectFit", "classNames", "onClick", "width", "height", "borderless"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import { Image as AntImage, Skeleton } from 'antd';
import { useResponsive, useThemeMode } from 'antd-style';
import { Eye } from 'lucide-react';
import { memo } from 'react';
import { Flexbox } from 'react-layout-kit';
import Icon from "../Icon";
import { useStyles } from "./style";
import usePreview from "./usePreview";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var Image = /*#__PURE__*/memo(function (_ref) {
  var wrapperClassName = _ref.wrapperClassName,
    style = _ref.style,
    preview = _ref.preview,
    isLoading = _ref.isLoading,
    _ref$minSize = _ref.minSize,
    minSize = _ref$minSize === void 0 ? 64 : _ref$minSize,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? '100%' : _ref$size,
    actions = _ref.actions,
    alwaysShowActions = _ref.alwaysShowActions,
    _ref$objectFit = _ref.objectFit,
    objectFit = _ref$objectFit === void 0 ? 'cover' : _ref$objectFit,
    _ref$classNames = _ref.classNames,
    classNames = _ref$classNames === void 0 ? {} : _ref$classNames,
    onClick = _ref.onClick,
    width = _ref.width,
    height = _ref.height,
    borderless = _ref.borderless,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useResponsive = useResponsive(),
    mobile = _useResponsive.mobile;
  var _useThemeMode = useThemeMode(),
    isDarkMode = _useThemeMode.isDarkMode;
  var _useStyles = useStyles({
      alwaysShowActions: alwaysShowActions,
      borderless: borderless,
      minSize: minSize,
      objectFit: objectFit,
      size: size
    }),
    styles = _useStyles.styles,
    cx = _useStyles.cx,
    theme = _useStyles.theme;
  var mergePreivew = usePreview(preview);
  if (isLoading) return /*#__PURE__*/_jsx("div", {
    onClick: onClick,
    children: /*#__PURE__*/_jsx(Skeleton.Avatar, {
      active: true,
      style: {
        borderRadius: theme.borderRadius,
        height: height,
        maxHeight: size,
        maxWidth: size,
        minHeight: minSize,
        minWidth: minSize,
        width: width
      }
    })
  });
  return /*#__PURE__*/_jsxs(Flexbox, {
    className: cx(styles.imageWrapper, wrapperClassName),
    style: style,
    children: [actions && /*#__PURE__*/_jsx("div", {
      className: styles.actions,
      children: actions
    }), /*#__PURE__*/_jsx(AntImage, _objectSpread({
      className: classNames.image,
      fallback: isDarkMode ? 'https://gw.alipayobjects.com/zos/kitchen/nhzBb%24r0Cm/image_off_dark.webp' : 'https://gw.alipayobjects.com/zos/kitchen/QAvkgt30Ys/image_off_light.webp',
      height: height,
      loading: 'lazy',
      onClick: onClick,
      preview: preview === false ? false : _objectSpread({
        mask: mobile ? false : actions ? /*#__PURE__*/_jsx("div", {}) : /*#__PURE__*/_jsx(Icon, {
          icon: Eye,
          size: 'normal'
        })
      }, mergePreivew),
      width: width,
      wrapperClassName: cx(styles.image, classNames.wrapper)
    }, rest))]
  });
});
export default Image;
