// @lobehub/ui/es/Avatar/index.js
'use client';

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var _excluded = ["className", "avatar", "title", "animation", "size", "shape", "background", "style", "unoptimized", "alt"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import { getEmoji } from '@lobehub/fluent-emoji';
import { Avatar as AntAvatar } from 'antd';
import { forwardRef, isValidElement, useMemo } from 'react';
import FluentEmoji from "../FluentEmoji";
import Img from "../Img";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
var Avatar = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var className = _ref.className,
    avatar = _ref.avatar,
    title = _ref.title,
    animation = _ref.animation,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 40 : _ref$size,
    _ref$shape = _ref.shape,
    shape = _ref$shape === void 0 ? 'circle' : _ref$shape,
    _ref$background = _ref.background,
    background = _ref$background === void 0 ? 'rgba(0,0,0,0)' : _ref$background,
    style = _ref.style,
    unoptimized = _ref.unoptimized,
    alt = _ref.alt,
    rest = _objectWithoutProperties(_ref, _excluded);
  var isStringAvatar = typeof avatar === 'string';
  var isDefaultAntAvatar = Boolean(avatar && (['/', 'http', 'data:'].some(function (index) {
    return isStringAvatar && avatar.startsWith(index);
  }) || /*#__PURE__*/isValidElement(avatar)));
  var emoji = useMemo(function () {
    return avatar && !isDefaultAntAvatar && isStringAvatar && getEmoji(avatar);
  }, [avatar]);
  var _useStyles = useStyles({
      background: background,
      isEmoji: Boolean(emoji),
      size: size
    }),
    styles = _useStyles.styles,
    cx = _useStyles.cx;
  var text = String(isDefaultAntAvatar ? title : avatar);
  var avatarProps = _objectSpread({
    alt: alt || title || 'avatar',
    className: cx(styles.avatar, className),
    shape: shape,
    size: size,
    style: rest !== null && rest !== void 0 && rest.onClick ? style : _objectSpread({
      cursor: 'default'
    }, style)
  }, rest);
  if (isDefaultAntAvatar) return /*#__PURE__*/_jsx(AntAvatar, _objectSpread({
    ref: ref,
    src: typeof avatar === 'string' ? /*#__PURE__*/_jsx(Img, {
      alt: avatarProps.alt,
      height: size,
      loading: 'lazy',
      src: avatar,
      unoptimized: unoptimized,
      width: size
    }) : avatar
  }, avatarProps));
  return /*#__PURE__*/_jsx(AntAvatar, _objectSpread(_objectSpread({
    ref: ref
  }, avatarProps), {}, {
    children: emoji ? /*#__PURE__*/_jsx(FluentEmoji, {
      emoji: emoji,
      size: size * 0.8,
      type: animation ? 'anim' : '3d',
      unoptimized: unoptimized
    }) : text === null || text === void 0 ? void 0 : text.toUpperCase().slice(0, 2)
  }));
});
export default Avatar;
