// @lobehub/ui/es/Mermaid/components/MermaidContainer.js
import { kebabCase } from 'lodash-es';
import { Loader2 } from 'lucide-react';
import { memo, useId } from 'react';
import { Center, Flexbox } from 'react-layout-kit';
import Icon from "../../Icon";
import { useMermaid } from "../../hooks/useMermaid";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var MermaidContainer = /*#__PURE__*/memo(function (_ref) {
  var _ref$children = _ref.children,
    children = _ref$children === void 0 ? '' : _ref$children;
  var _useStyles = useStyles(),
    styles = _useStyles.styles;
  var id = useId();
  var mermaidId = kebabCase("mermaid-".concat(id));
  var _useMermaid = useMermaid(mermaidId, children),
    data = _useMermaid.data,
    isLoading = _useMermaid.isLoading;
  if (!data) return null;
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx(Center, {
      as: 'pre',
      dangerouslySetInnerHTML: {
        __html: data
      },
      id: "mermaid-".concat(id)
    }), isLoading && /*#__PURE__*/_jsx(Flexbox, {
      align: 'center',
      className: styles.loading,
      gap: 8,
      horizontal: true,
      justify: 'center',
      children: /*#__PURE__*/_jsx(Icon, {
        icon: Loader2,
        spin: true
      })
    })]
  });
});
export default MermaidContainer;
