// @lobehub/tts/es/react/useSpeechSynthes/index.js
var _excluded = ["voice", "rate", "pitch"];
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import { useCallback, useEffect, useMemo, useState } from 'react';
import { SpeechSynthesis, SpeechSynthesisUtterance } from "../../core/const/polyfill";
export var useSpeechSynthes = function useSpeechSynthes(defaultText, _ref) {
  var voice = _ref.voice,
    rate = _ref.rate,
    pitch = _ref.pitch,
    options = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState(SpeechSynthesis === null || SpeechSynthesis === void 0 ? void 0 : SpeechSynthesis.getVoices()),
    _useState2 = _slicedToArray(_useState, 2),
    voiceList = _useState2[0],
    setVoiceList = _useState2[1];
  var _useState3 = useState(defaultText),
    _useState4 = _slicedToArray(_useState3, 2),
    text = _useState4[0],
    setText = _useState4[1];
  var _useState5 = useState(false),
    _useState6 = _slicedToArray(_useState5, 2),
    isLoading = _useState6[0],
    setIsLoading = _useState6[1];
  var speechSynthesisUtterance = useMemo(function () {
    if (!SpeechSynthesisUtterance) return;
    var utterance = new SpeechSynthesisUtterance(text);
    utterance.voice = voiceList.find(function (item) {
      return item.name === voice;
    });
    utterance.onstart = function () {
      return setIsLoading(true);
    };
    utterance.onend = function () {
      return setIsLoading(false);
    };
    if (pitch) utterance.pitch = pitch * 10;
    if (rate) utterance.rate = rate * 10;
    return utterance;
  }, [text, voiceList, rate, pitch, voice]);
  useEffect(function () {
    if (!SpeechSynthesis) return;
    SpeechSynthesis.onvoiceschanged = function () {
      setVoiceList(SpeechSynthesis === null || SpeechSynthesis === void 0 ? void 0 : SpeechSynthesis.getVoices());
    };
    SpeechSynthesis.onstart = function () {
      return setIsLoading(true);
    };
    SpeechSynthesis.onend = function () {
      return setIsLoading(false);
    };
  }, []);
  var handleStart = useCallback(function () {
    var _options$onStart;
    options === null || options === void 0 || (_options$onStart = options.onStart) === null || _options$onStart === void 0 || _options$onStart.call(options);
    SpeechSynthesis === null || SpeechSynthesis === void 0 || SpeechSynthesis.speak(speechSynthesisUtterance);
  }, [speechSynthesisUtterance]);
  var handleStop = useCallback(function () {
    var _options$onStop, _speechSynthesis;
    options === null || options === void 0 || (_options$onStop = options.onStop) === null || _options$onStop === void 0 || _options$onStop.call(options);
    (_speechSynthesis = speechSynthesis) === null || _speechSynthesis === void 0 || _speechSynthesis.cancel();
    setIsLoading(false);
  }, []);
  return {
    isLoading: isLoading,
    setText: setText,
    start: handleStart,
    stop: handleStop
  };
};
